File: devtools/sdk/dart/lib/src/connection.dart
import 'dart:async';
import 'dart:convert';
import 'dart:collection';
import 'dart:io';

/// TCP JSON-line client that connects to the devtools source server.
///
/// Auto-reconnects every 5 seconds and buffers up to 100 messages when
/// disconnected. All operations are non-blocking and safe to call from
/// any isolate.
class Connection {
  final String host;
  final int port;
  final String appName;

  static const int _maxBufferSize = 100;
  static const Duration _reconnectDelay = Duration(seconds: 5);
  static const Duration _connectTimeout = Duration(seconds: 3);

  Socket? _socket;
  bool _connected = false;
  bool _closed = false;
  final Queue<String> _buffer = Queue<String>();
  Timer? _reconnectTimer;

  Connection(this.host, this.port, this.appName) {
    _connect();
  }

  Future<void> _connect() async {
    if (_closed) return;

    try {
      final socket = await Socket.connect(
        host,
        port,
        timeout: _connectTimeout,
      );

      if (_closed) {
        socket.destroy();
        return;
      }

      _socket = socket;
      _connected = true;

      // Flush buffered messages
      _flushBuffer();

      // Listen for disconnect
      socket.listen(
        (_) {}, // Discard incoming data; we're send-only
        onError: (_) => _onDisconnect(socket),
        onDone: () => _onDisconnect(socket),
        cancelOnError: true,
      );
    } catch (_) {
      // Connection failed; schedule reconnect
      _scheduleReconnect();
    }
  }

  void _onDisconnect(Socket socket) {
    if (_socket == socket) {
      _connected = false;
      _socket = null;
      try {
        socket.destroy();
      } catch (_) {}
    }
    _scheduleReconnect();
  }

  void _scheduleReconnect() {
    if (_closed || _reconnectTimer != null) return;
    _reconnectTimer = Timer(_reconnectDelay, () {
      _reconnectTimer = null;
      if (!_closed && !_connected) {
        _connect();
      }
    });
  }

  void _flushBuffer() {
    while (_buffer.isNotEmpty && _connected && _socket != null) {
      final msg = _buffer.first;
      try {
        _socket!.write(msg);
        _buffer.removeFirst();
      } catch (_) {
        break;
      }
    }
  }

  /// Send an event to the devtools server. Non-blocking.
  void send(String type, Map<String, dynamic> data) {
    if (_closed) return;

    final event = {
      'type': type,
      'data': data,
      'source': appName,
      'runtime': 'dart',
      'timestamp': DateTime.now().millisecondsSinceEpoch,
    };

    String msg;
    try {
      msg = '${jsonEncode(event)}\n';
    } catch (_) {
      return;
    }

    if (_connected && _socket != null) {
      try {
        _socket!.write(msg);
        return;
      } catch (_) {
        // Write failed; fall through to buffer
        _connected = false;
        try {
          _socket?.destroy();
        } catch (_) {}
        _socket = null;
        _scheduleReconnect();
      }
    }

    // Buffer up to _maxBufferSize messages
    if (_buffer.length < _maxBufferSize) {
      _buffer.add(msg);
    }
  }

  /// Close the connection and stop reconnection attempts.
  void close() {
    if (_closed) return;
    _closed = true;
    _connected = false;

    _reconnectTimer?.cancel();
    _reconnectTimer = null;

    try {
      _socket?.destroy();
    } catch (_) {}
    _socket = null;

    _buffer.clear();
  }
}

File: devtools/sdk/go/connection.go
package cloudmock

import (
	"encoding/json"
	"fmt"
	"net"
	"sync"
	"time"
)

const (
	maxBufferSize    = 100
	reconnectDelay   = 5 * time.Second
	connectTimeout   = 3 * time.Second
	writeTimeout     = 2 * time.Second
)

// event is the wire format sent to the devtools source server.
type event struct {
	Type      string         `json:"type"`
	Data      map[string]any `json:"data"`
	Source    string         `json:"source"`
	Runtime   string         `json:"runtime"`
	Timestamp int64          `json:"timestamp"`
}

// connection is a TCP JSON-line client that connects to the devtools source server.
// It auto-reconnects every 5 seconds and buffers up to 100 messages when disconnected.
// All operations are non-blocking and goroutine-safe.
type connection struct {
	host    string
	port    int
	appName string

	mu        sync.Mutex
	conn      net.Conn
	connected bool
	closed    bool
	buffer    []string

	done chan struct{}
}

func newConnection(host string, port int, appName string) *connection {
	c := &connection{
		host:    host,
		port:    port,
		appName: appName,
		buffer:  make([]string, 0, maxBufferSize),
		done:    make(chan struct{}),
	}
	go c.connectLoop()
	return c
}

func (c *connection) connectLoop() {
	// Attempt first connection immediately
	c.tryConnect()

	for {
		select {
		case <-c.done:
			return
		case <-time.After(reconnectDelay):
			c.mu.Lock()
			needsConnect := !c.connected && !c.closed
			c.mu.Unlock()
			if needsConnect {
				c.tryConnect()
			}
		}
	}
}

func (c *connection) tryConnect() {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return
	}
	c.mu.Unlock()

	addr := net.JoinHostPort(c.host, fmt.Sprintf("%d", c.port))
	conn, err := net.DialTimeout("tcp", addr, connectTimeout)
	if err != nil {
		return
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		conn.Close()
		return
	}

	c.conn = conn
	c.connected = true

	// Flush buffered messages
	remaining := make([]string, 0)
	for _, msg := range c.buffer {
		if err := c.writeLocked(msg); err != nil {
			remaining = append(remaining, msg)
			break
		}
	}
	c.buffer = remaining

	// Start a goroutine to detect disconnection by reading
	go c.readUntilClose(conn)
}

func (c *connection) readUntilClose(conn net.Conn) {
	buf := make([]byte, 256)
	for {
		_, err := conn.Read(buf)
		if err != nil {
			c.mu.Lock()
			if c.conn == conn {
				c.connected = false
				c.conn = nil
				conn.Close()
			}
			c.mu.Unlock()
			return
		}
	}
}

// send serializes and sends an event to the devtools server.
// Non-blocking: buffers the message if not connected.
func (c *connection) send(typ string, data map[string]any) {
	e := event{
		Type:      typ,
		Data:      data,
		Source:    c.appName,
		Runtime:   "go",
		Timestamp: time.Now().UnixMilli(),
	}

	b, err := json.Marshal(e)
	if err != nil {
		return
	}
	msg := string(b) + "\n"

	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return
	}

	if c.connected && c.conn != nil {
		if err := c.writeLocked(msg); err == nil {
			return
		}
		// Write failed; fall through to buffer
		c.connected = false
		if c.conn != nil {
			c.conn.Close()
			c.conn = nil
		}
	}

	// Buffer up to maxBufferSize messages
	if len(c.buffer) < maxBufferSize {
		c.buffer = append(c.buffer, msg)
	}
}

// writeLocked writes a message to the TCP connection. Must be called with mu held.
func (c *connection) writeLocked(msg string) error {
	if c.conn == nil {
		return fmt.Errorf("no connection")
	}
	c.conn.SetWriteDeadline(time.Now().Add(writeTimeout))
	_, err := c.conn.Write([]byte(msg))
	return err
}

// close shuts down the connection and stops reconnection attempts.
func (c *connection) close() {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return
	}
	c.closed = true
	close(c.done)

	if c.conn != nil {
		c.conn.Close()
		c.conn = nil
	}
	c.connected = false
	c.buffer = nil
}
